reports/macbox7.md

```diff
@@ -0,0 +1,307 @@
+  - link: https://gitlab.com/coala/bears/coala-pyflakes/commit/87dddfa0709591d2925df849fc16f542d08e8e88
+  - details: requirements.txt: Use bears master
+- 50:
+  - repo: pyflakes
+  - link: https://github.com/PyCQA/pyflakes/commit/9dd73ec54411a563410872b47e76f0f89f34ecfe
+  - details: Handle subtrees as input to checker
+- 51:
+  - repo: pyflakes
+  - link: https://github.com/PyCQA/pyflakes/commit/fb1f223ac2141772f140be6f1108d02009c1d658
+  - details: checker.py: Handle UnboundLocal error for builtins
+- 52:
+  - repo: pyflakes
+  - link: https://github.com/PyCQA/pyflakes/pull/329
+  - details: checker.py: Check for invalid print syntax
+---
+
+### Integrate Pyflakes-Enhanced-AST into coala
+
+#### Work Done
+
+There has not been an option to use the pyflakes-enhanced-AST based
+checker API to create custom linters. Thus, the full potential of
+enhanced AST isn't utilized, a whole lot of rework is required to do
+the basic traversing and collection of important nodes. Pyflakes provides
+with a basic API that does the traversing. So, if a developer uses enhanced
+AST he just needs to work on the implementation of the new logic that his/her
+plugin provides and not about the fidelity of the basic node handlers.
+
+**Major tasks of the project:**
+
+1. Create a separate repository `coala-pyflakes` to house all
+   pyflakes based plugins.
+2. Create package `pyflakes_bears` to hold `PyFlakesASTBear` and all the
+   coala bears developed using it.
+3. Create package `pyflakes_generic_plugins` that consists generic plugins
+   which can be integrated with `flake8` and other such similar wrappers.
+4. Design and develop `PyFlakesASTBear` that provides helper functions
+   that can be used to traverse pyflakes AST of the input python module.
+5. Establish the proof of concept by creating `NoFutureImportBear` that
+   detects and fixes any usage of `__future__` imports in a python file.
+6. Add handlers for `ARGS` and `ARGUMENTS` and provide the functionality
+   of partial code analysis in `pyflakes`.
+7. Create a `PEP8NamingBear` that detects all pep8 naming violations
+   in a python file. Special handling is also done to detect violations
+   in doctests which other similar plugins fail to perform.
+8. Create a `BuiltinsBear` that is capable of detecting redefinition
+   of python builtins in a file.
+
+
+#### Challenges
+
+The challenge of the project was to implement the `PyFlakesASTBear` in
+such a way that a plugin developer need not worry about writing his own
+traversing logic and need to have much no more than the knowledge of
+PyFlakesASTBear nodes to write the bears. However, designing the logic
+first by writing cEP(coala enhancement proposal) for
+[coala-pyflakes](https://github.com/coala/cEPs/blob/master/cEP-0025.md) helped
+me a lot to understand the problem to a deep level and come out with a logic
+to tackle it.
+
+#### Future prospects of the projects
+
+With `PyFlakesASTBear` ready bear developers can use it to detect codesmells
```

Comment:
"ready" isnt clear.  Is it ready now?  If so, using "with .. ready" is redundant, at best.

Also migrating flake8 plugins to this new infrastructure should be easy tasks that people can now work on.

And more importantly, we had planned for this bear framework to be able to load & run flake8 ast plugins as-is, so porting wasnt necessary; that can be mentioned as a medium to large sized project which is now feasible.